### PR TITLE
Implement comparison for pre-release tags.

### LINKED ISSUE #53: Implement comparison rules for pre-release tags
The comparison functions in `version_req.rs` currently ignore prerelease tags.  They should instead follow the [NPM rules for prerelease tags](https://docs.npmjs.com/misc/semver#prerelease-tags):

> If a version has a prerelease tag (for example, `1.2.3-alpha.3`) then it will only be allowed to satisfy comparator sets if at least one comparator with the same `[major, minor, patch]` tuple also has a prerelease tag.

### PR DESCRIPTION
**Pull request by mbrubeck**
This fixes #53 by implementing the rules specified at https://docs.npmjs.com/misc/semver#prerelease-tags

r? @steveklabnik

### CODE REVIEW

**Comment by steveklabnik** [comment]
:confetti_ball: 

thanks!

**Comment by globin** [comment]
:+1:

**Comment by steveklabnik** [comment]
So, I cannot figure out why, but I reverted this in #71 because it causes error's in cargo's tests like this:

https://gist.github.com/steveklabnik/3837db51cdaf2c61e5b1

I am confused because, with this patch applied, if I add:

```diff
         let r = req("*");
-        assert_match(&r, &["0.9.1", "2.9.0", "0.0.9", "1.0.1", "1.1.1"]);
+        assert_match(&r, &["0.9.1", "2.9.0", "0.0.9", "1.0.1", "1.1.1", "0.0.1"]);
         assert_not_match(&r, &[]);
```

it does pass. So I'm not sure what's up here. Any ideas @mbrubeck ?

**Comment by steveklabnik** [comment]
I am suspecting this is a bug in cargo, it seems that it says it matches: https://gist.github.com/steveklabnik/e9bc30f81158234b81f6

File: src/version_req.rs [hunk 1]
```diff
@@ -232,7 +245,8 @@ impl VersionReq {
     /// assert!(exact.matches(&version));
     /// ```
     pub fn matches(&self, version: &Version) -> bool {
-        self.predicates.iter().all(|p| p.matches(version))
+        self.predicates.iter().all(|p| p.matches(version)) &&
+            self.predicates.iter().any(|p| p.pre_tag_is_compatible(version))
```
Comment (steveklabnik) [inline src/version_req.rs]: ah ha! no, I think this is the bug. 

```
fn main() {
    let v: Vec<i32> = vec![];
    
    println!("any {}", v.iter().any(|_| true));
    println!("all {}", v.iter().all(|_| true));
}
```

prints `false` and then `true`, in other words, when `predicates` is empty, this is wrong.

Comment (steveklabnik) [inline src/version_req.rs]: Yup!

```patch
diff --git a/src/version_req.rs b/src/version_req.rs
index 19e7252..620ccb7 100644
--- a/src/version_req.rs
+++ b/src/version_req.rs
@@ -253,6 +253,10 @@ impl VersionReq {
         println!("pre_compat:: {}", pre_compat);
 
         println!("predicates: {:?}", self.predicates);
+        if self.predicates.len() == 0 {
+            return true;
+        }
+
         self.predicates.iter().all(|p| p.matches(version)) &&
             self.predicates.iter().any(|p| p.pre_tag_is_compatible(version))
     }

```

this fixes Cargo while passing our test suite.

Comment (steveklabnik) [inline src/version_req.rs]: actually, it fixes one instance, there's probably more

**Pull request merged**